fdbserver/Resolver.actor.cpp

```diff
@@ -18,36 +18,72 @@
  * limitations under the License.
  */
 
-#include "flow/ActorCollection.h"
+#include <chrono>
+
 #include "fdbclient/NativeAPI.actor.h"
-#include "fdbserver/ResolverInterface.h"
-#include "fdbserver/MasterInterface.h"
-#include "fdbserver/WorkerInterface.actor.h"
-#include "fdbserver/WaitFailure.h"
+#include "fdbclient/SystemData.h"
+#include "fdbserver/ConflictSet.h"
 #include "fdbserver/Knobs.h"
-#include "fdbserver/ServerDBInfo.h"
+#include "fdbserver/MasterInterface.h"
 #include "fdbserver/Orderer.actor.h"
-#include "fdbserver/ConflictSet.h"
+#include "fdbserver/PartMerger.h"
+#include "fdbserver/ResolverInterface.h"
+#include "fdbserver/ServerDBInfo.h"
 #include "fdbserver/StorageMetrics.h"
-#include "fdbclient/SystemData.h"
+#include "fdbserver/TimedKVCache.h"
+#include "fdbserver/WaitFailure.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "flow/ActorCollection.h"
+
+#include "debug.h"
+
 #include "flow/actorcompiler.h"  // This must be the last #include.
 
 namespace {
+
+class ResolveTransactionBatchRequestMerger : public PartMerger<UID, ResolveTransactionBatchRequest> {
+protected:
+	virtual void merge(value_t& value, const value_t& incomingRequest) override {
+		ASSERT(incomingRequest.transactions.size() == 1);
+
+		auto& arena = value.arena;
+		auto& current = value.transactions[0];
+		auto& incoming = incomingRequest.transactions[0];
+
+		ASSERT(current.read_snapshot == incoming.read_snapshot);
+		ASSERT(current.report_conflicting_keys == incoming.report_conflicting_keys);
+
+		// FIXME is this efficient?
+		current.read_conflict_ranges.append(arena, incoming.read_conflict_ranges.begin(),
+		                                    incoming.read_conflict_ranges.size());
+		current.write_conflict_ranges.append(arena, incoming.write_conflict_ranges.begin(),
+		                                     incoming.write_conflict_ranges.size());
```

Comment:
If you use `append`, then you also need to call `arena.dependsOn(incoming.arena)` so that `value` can hold a reference to the `incoming.arena`. Otherwise, if `incoming` is destroyed, the `Ref`s would point to invalid memory.

`append_deep` does the deep copy, but is inefficient here. Because `append` copy `write_conflict_ranges` by reference only.

Reply:
Regarding to this, the `arena` will not be invalidated in this particular situation since the other arenas belongs to requests that are still waiting for the final Promise gets resolved. But my other two concerns:
 - Is there a better approach? Is it possible to merge arenas?
 - Even the conflicts are copied by reference, is it possible that the current arena get out-of-space? If so, how do I avoid this?

Reply:
`arena.dependsOn(incoming.arena)` is the way to merge two arenas, but copy of `VectorRef` is still needed. One way to avoid copy is probably use a vector of `pair<Arena, VectorRef>`, since both `Arean` and `VectorRef` copy is by reference.

Generally, the Arena should outlive its usage. That's why the object owns the `Arena`.